### PR TITLE
fix: add hook to handle incomplete commands

### PR DESCRIPTION
**Pull request by peternhale**
### What does this PR do?
Adds hook to handle incomplete commands for tax-free
### What issues does this PR fix or reference?
@W-12212675@

### CODE REVIEW

**Review by WillieRuemmele — APPROVED** [review]

**Review by mshanemc — APPROVED** [review]

**Review by mshanemc — APPROVED** [review]
code approved, please make sure this gets QA'd thoroughly

**Comment by cristiand391** [comment]
# QA notes:

✅ 1,2,3,4,5 args, shows all possible matches:
```
➜  sfdx-cli git:(phale/command-incompolete) ./bin/run install
? Which of these commands do you mean (Use arrow keys)
❯ plugins install
  force package beta install
  force package beta install report
  force package install
  force package install report
```

manually checked options returned against what `commands` returns by running:
`./bin/run commands --no-header --columns 'Command' | fzf`

❌  typed partial command, the hook finds 1 match and runs what it found, even though I didn't type the full command nor answered (no prompt rendered):

```
➜  sfdx-cli git:(phale/command-incompolete) ./bin/run verify -h
Validate a digital signature.

USAGE
  $ sfdx plugins trust verify -n <value> [--json] [-r <value>]

FLAGS
  -n, --npm=<value>       (required) Specify the npm name. This can include a tag/version.
  -r, --registry=<value>  The registry name. The behavior is the same as npm.

GLOBAL FLAGS
  --json  Format output as json.

DESCRIPTION
  Validate a digital signature.

  Verifies the digital signature on an npm package matches the signature and key stored at the expected URLs.

EXAMPLES
  $ sfdx plugins trust verify --npm @scope/npmName --registry http://my.repo.org:4874

  $ sfdx plugins trust verify --npm @scope/npmName

➜  sfdx-cli git:(phale/command-incompolete) ./bin/run which verify
 ›   Error: command verify not found
➜  sfdx-cli git:(phale/command-incompolete) ./bin/run search
? Search for a command verify
❯ plugins trust verify
──────────────
Validate a digital signature.
```

Users might run `sfdx verify` and think it's a valid command, but then running `sfdx which <command>` will fail because it doesn't resolve commands the same way the hook does.

This seems to happen when you type a word that is part of **only** 1 command, a few examples:

`plugins trust verify`: the hook will run that command whenever you try `./bin/run trust` or `./bin/run verify`.
`auth jwt grant`: `./bin/run auth` shows all `auth *` cmds, `./bin/run jwt` executes`auth jwt grant`.
`autocomplete script`: `./bin/run script` executes`autocomplete script`.

⚠️   typed partial command, expected prompt asking all possible options but the hook ran the first match:
```
➜  sfdx-cli git:(phale/command-incompolete) ./bin/run plugins
No plugins installed.
➜  sfdx-cli git:(phale/command-incompolete) ./bin/run search
? Search for a command plugins
❯ plugins
  plugins inspect
  plugins install
  plugins add
  plugins link
  plugins uninstall
  plugins unlink
  plugins remove
  plugins update
  plugins generate
(Move up and down to reveal more choices)
➜  sfdx-cli git:(phale/command-incompolete)
```
This happens when the typed arg is an actual command, I think this is ok as the command is correct but also was expecting the prompt with what `./bin/run search` shows for `plugins`. 🤷🏼 

Another example:
the following are valid commands:

1. `force package beta version create`
2. `force package beta version create list`
3. `force package beta version create report`

`./bin/run force package create`: executes `force package beta version create`, no prompt.

✅ typed an invalid command, oclif fired the `command_not_found` instead of `command_incomplete`:
```
➜  sfdx-cli git:(phale/command-incompolete) ✗ ./bin/run asfsaf
 ›   Warning: asfsaf is not a sfdx command.
Did you mean alias set? [y/n]: n
 ›   Error: Run sfdx help for a list of available commands.
```

see: https://github.com/oclif/core/blob/main/src/config/config.ts#L331-L333

❌ the hook doesn't respect the short flag for help, always shows the full help text:
```
➜  sfdx-cli git:(phale/command-incompolete) ✗ ./bin/run org list -h
? Which of these commands do you mean force org list
list all orgs you’ve created or authenticated to

USAGE
  $ sfdx force org list [--all] [-p --clean] [--skipconnectionstatus] [--verbose] [--json] [--loglevel
    trace|debug|info|warn|error|fatal|TRACE|DEBUG|INFO|WARN|ERROR|FATAL]

FLAGS
  -p, --noprompt                                                                    do not prompt for confirmation
  --all                                                                             include expired, deleted, and unknown-status scratch orgs
  --clean                                                                           remove all local org authorizations for non-active scratch orgs.  Use
                                                                                    auth:logout to remove non-scratch orgs
  --json                                                                            format output as json
  --loglevel=(trace|debug|info|warn|error|fatal|TRACE|DEBUG|INFO|WARN|ERROR|FATAL)  [default: warn] logging level for this command invocation
  --skipconnectionstatus                                                            skip retrieving the connection status of non-scratch orgs
  --verbose                                                                         list more information about each org

DESCRIPTION
  list all orgs you’ve created or authenticated to

EXAMPLES
  $ sfdx force:org:list

  $ sfdx force:org:list --verbose --json

  $ sfdx force:org:list --verbose --json > tmp/MyOrgList.json

➜  sfdx-cli git:(phale/command-incompolete) ✗
```
compare it with what `sfdx force org list -h` returns.

✅ prompt times out after 10s:
```
➜  sfdx-cli git:(phale/command-incompolete) ✗ ./bin/run force list
? Which of these commands do you mean (Use arrow keys)
❯ force community template list
  alias list
  force apex log list
  auth list
  config list
  force org list
  force schema sobject list
(Move up and down to reveal more choices)
    Error: Timed out after 10000 ms.
➜  sfdx-cli git:(phale/command-incompolete) ✗ echo $?
1
```

⚠️ the prompt is still rendered even if the process isn't attached to a terminal:
```
➜  sfdx-cli git:(phale/command-incompolete) ✗ ./bin/run force list | node -p '`Attached to a terminal?: ${Boolean(process.stdin.isTTY)}`'
Attached to a terminal?: false
    Error: Timed out after 10000 ms.
➜  sfdx-cli git:(phale/command-incompolete) ✗ sfdx force list | node -p '`Attached to a terminal?: ${Boolean(process.stdin.isTTY)}`'
Attached to a terminal?: false
 ›   Error: command force:list not found
➜  sfdx-cli git:(phale/command-incompolete) ✗
```

`./bin/run ...` still timed out after 10s so it will not block a CI process if there's a typo in a pipeline.
I think this is fine (we could also save those 10s by skipping the prompt is `process.stdin.isTTY === false`. If you want you can specify the 10s timeout in the hook rather than relying on the default timeout set in `sf-plugins-core` in the case we change it.

File: src/hooks/incomplete.ts [hunk 1]
```diff
@@ -30,7 +33,17 @@ const hook: Hook.CommandIncomplete = async function ({ config, matches, argv })
     return config.runCommand('help', [toStandardizedId(command, config)]);
   }
 
-  return config.runCommand(toStandardizedId(command, config), argv);
+  const standardCommand = toStandardizedId(command, config);
```
Comment (cristiand391) [inline src/hooks/incomplete.ts:36]: nit: use `toConfigureId` to respect the topicSeparator (currently: ' ') as `determineCommand` does.

**Review by cristiand391 — APPROVED** [review]

**Review by cristiand391 — APPROVED** [review]

**Comment by cristiand391** [comment]
qa notes:

✅ emits warning there's just one match

Screen Shot 2022-12-30 at 10 31 01

**Pull request merged**